### PR TITLE
WGSL: Allow imports anywhere

### PR DESCRIPTION
**Pull request by DJMcNab**
Permit using an import anywhere within the file.

Note that only certain locations are supported by wgsl-analyzer. However, this is better than requiring them also to be on newlines. If you need a new place, we can probably add it to wgsl-analyzer.

We intentionally only support `ifdef` etc. at the start of a line, to avoid having to think too hard about it.

The preprocessor also doesn't have any tests, so if something here looks wrong it probably is.

### CODE REVIEW

File: piet-wgsl/src/shaders/preprocess.rs [hunk 1]
```diff
@@ -71,29 +87,53 @@ pub fn preprocess(
...
+                        directive_end.find(|c: char| !c.is_whitespace())
+                    {
+                        import_name_start
+                    } else {
+                        eprintln!("#import needs a non_whitespace argument (line {line_number})");
+                        continue 'outer;
+                    };
+                    let import_name_start = &directive_end[import_name_start..];
+                    let import_name_end_index = import_name_start
+                        .find(|c| !(matches!(c, '_') || c.is_alphanumeric()))
```
Comment (raphlinus) [inline piet-wgsl/src/shaders/preprocess.rs:115]: Any special reason `matches!` rather than just `c == '_'`? Looking at the code I'm surprised this compiles without `c` having an explicit type, maybe matches is forcing type equality while PartialEq would be looser.

File: piet-wgsl/src/shaders/preprocess.rs [hunk 2]
```diff
@@ -41,25 +41,41 @@ pub fn preprocess(
 ) -> String {
     let mut output = String::with_capacity(input.len());
     let mut stack = vec![];
-
-    for (line_number, line) in input.lines().enumerate() {
-        let trimmed = line.trim();
-        if trimmed.starts_with("#") {
-            let trimmed = &trimmed[1..];
-            let val_idx = trimmed
+    'outer: for (line_number, mut line) in input.lines().enumerate() {
```
Comment (raphlinus) [inline piet-wgsl/src/shaders/preprocess.rs:44]: Oooh, fancy, labeled loops. I found the flow a little hard to follow, but don't have a suggestion how to improve it.

File: piet-wgsl/src/shaders/preprocess.rs [hunk 3]
```diff
@@ -71,29 +87,53 @@ pub fn preprocess(
...
+                    eprintln!("Unknown preprocessor directive `{val}` (line {line_number})");
                 }
             }
-        } else {
-            if stack.last().map(|x| x.active).unwrap_or(true) {
-                output.push_str(line);
-                output.push('\n');
-            }
+        }
+        if stack.last().map(|x| x.active).unwrap_or(true) {
```
Comment (raphlinus) [inline piet-wgsl/src/shaders/preprocess.rs:134]: Is this actually correct? It seems like a sequence of `#ifdef resolves_false`, `#ifdef resolves_true`, `inner`, `#endif`, `#endif` will actually cause inner to be included. My intuition (without having experimented) is that we want an `all` on the `active` field of all entries on the stack.

File: piet-wgsl/src/shaders/preprocess.rs [hunk 4]
```diff
@@ -71,29 +87,53 @@ pub fn preprocess(
...
+                    let import_name_start = if let Some(import_name_start) =
+                        directive_end.find(|c: char| !c.is_whitespace())
+                    {
+                        import_name_start
+                    } else {
+                        eprintln!("#import needs a non_whitespace argument (line {line_number})");
+                        continue 'outer;
+                    };
+                    let import_name_start = &directive_end[import_name_start..];
+                    let import_name_end_index = import_name_start
```
Comment (raphlinus) [inline piet-wgsl/src/shaders/preprocess.rs:114]: It seems to me this is doing the same logic as the calculation of directive_len but in a different way. I find this version slightly more intuitive and less fragile. Either way, unless there's a compelling reason it seems like the two should be the same.

**Review by raphlinus — APPROVED** [review]
Generally looks good but I had a bunch of nits :)

File: piet-wgsl/src/shaders/preprocess.rs (hunk 1 above)
Comment (DJMcNab) [reply piet-wgsl/src/shaders/preprocess.rs:115]: I think orginally there were several characters I wanted to allow there (e.g. `-`, I think)

It does also make the types work out, although obviously it's reasonable to be consistent.

File: piet-wgsl/src/shaders/preprocess.rs (hunk 2 above)
Comment (DJMcNab) [reply piet-wgsl/src/shaders/preprocess.rs:44]: Yeah. In https://github.com/linebender/piet-gpu/pull/196/commits/40f808b366b4a7087b4da97f756e480c5265c859 I tried a different way, but that was really messy.

File: piet-wgsl/src/shaders/preprocess.rs (hunk 3 above)
Comment (DJMcNab) [reply piet-wgsl/src/shaders/preprocess.rs:134]: Good point. Thanks

File: piet-wgsl/src/shaders/preprocess.rs (hunk 4 above)
Comment (DJMcNab) [reply piet-wgsl/src/shaders/preprocess.rs:114]: Well spotted, I've used the same technique for `directive_len ` now

**Comment by DJMcNab** [comment]
> Can we expect wgsl-analyzer to also be able to handle #import in the middle of a line?

In the sense that it's theoretically possible, yes. At the moment, the actual places are quite limited, but for any other places you need, adding such support should be *relatively* feasible

**Review by raphlinus — APPROVED** [review]
It occurs to me `all` is a monoid, so all the usual optimizations are possible. But I agree, let's optimize for code clarity, O(n) is fine if n is really small.

**Comment by DJMcNab** [comment]
> Hmm, that means actually using imports for constants will result in a red squiggle? 

Yes exactly, but if you poke me on the [zulip](https://xi.zulipchat.com/) I'd guess it should take maybe 30 minutes per syntactic location.

I can only think that we need some attributes support (potentially just top level attribute to avoid having to think about nesting). Apparently parameter declarations was also needed - that supports already exists, but I'm not sure why.

You should already be able to import global constants, which is equivalent to having it be in any other place you can refer to constants, just with slightly more indirection.

I think everything else should be equivalently constructable from these cases.

I'll merge now though
